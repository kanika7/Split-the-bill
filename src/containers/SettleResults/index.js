import React, { Component } from "react";
import PropTypes from "prop-types";
import { OWES, AVG_EXPENSE } from "../../constants";

import "./styles/index.scss";

class SettleResults extends Component {
  static debtArr = [];
  constructor(props) {
    super(props);
    this.state = {
      expenseArr: [],
      avgAmount: ""
    };
  }

  componentDidMount() {
    this.settleUp(this.props.expenses);
  }

  // Calculating average amount for each user.
  calculateAverage = entries => {
    const nameAndAmountSpent = {};
    const { totalAmount } = this.props;
    let totalUsers = 0;
    entries.forEach(entry => {
      if (!nameAndAmountSpent[entry.name]) {
        nameAndAmountSpent[entry.name] = entry.amount;
        totalUsers++;
      } else {
        nameAndAmountSpent[entry.name] += entry.amount;
      }
    });
    const perHeadAmount = totalAmount / totalUsers;
    return { nameAndAmountSpent, perHeadAmount };
  };

  // called for executing split of bills
  settleUp = entries => {
    const { perHeadAmount, nameAndAmountSpent } = this.calculateAverage(entries);
    const result = [];
    let str = "";

    for (let name in nameAndAmountSpent) {
      nameAndAmountSpent[name] -= perHeadAmount;
    }
    for (let i in nameAndAmountSpent) {
      if (nameAndAmountSpent[i] === 0) continue;
      // will only run for users who have to get money back
      if (nameAndAmountSpent[i] > 0) {
        for (let j in nameAndAmountSpent) {
          //will run for users who have to give the money to others
          if (nameAndAmountSpent[j] < 0) {
            if (nameAndAmountSpent[i] === -nameAndAmountSpent[j]) {
              str = `${j} ${OWES} ${i} ${nameAndAmountSpent[i].toFixed(2)}`;
              result.push(str);
              nameAndAmountSpent[i] = 0;
              nameAndAmountSpent[j] = 0;
            } else if (nameAndAmountSpent[i] > -nameAndAmountSpent[j]) {
              str = `${j} ${OWES} ${i} ${-nameAndAmountSpent[j].toFixed(2)}`;
              result.push(str);
              nameAndAmountSpent[i] -= nameAndAmountSpent[j];
              nameAndAmountSpent[j] = 0;
            } else {
              str = `${j} ${OWES} ${i} ${nameAndAmountSpent[i].toFixed(2)}`;
              result.push(str);
              nameAndAmountSpent[j] += nameAndAmountSpent[i];
              nameAndAmountSpent[i] = 0;
            }
          }
          if (nameAndAmountSpent[i] === 0) break;
        }
      }
    }

    this.setState({
      expenseArr: [...result],
      avgAmount: perHeadAmount
    });
  };

  render() {
    const { expenseArr, avgAmount } = this.state;

    return (
      <ul className="results-container">
        {avgAmount && (
          <li className="list-item">
            <span className="col-xs-12 col-lg-4">{`${AVG_EXPENSE} ${avgAmount.toFixed(2)}`}</span>
          </li>
        )}
        {expenseArr.map((exp, index) => (
          <li key={index} className="list-item">
            <span className="col-xs-12 col-lg-4">{exp}</span>
          </li>
        ))}
      </ul>
    );
  }
}

SettleResults.propTypes = {
  expenses: PropTypes.instanceOf(Array).isRequired,
  totalAmount: PropTypes.number.isRequired
};

export default SettleResults;
