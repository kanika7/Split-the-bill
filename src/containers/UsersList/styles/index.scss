@import "./default.scss";
@import "./desktop.scss";
