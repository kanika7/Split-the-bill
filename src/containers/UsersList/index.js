import React, { Component } from "react";
import SettleResults from "../SettleResults";
import { ADD_EXPENSE } from "../../constants";
import "./styles/index.scss";
import ExpenseList from "../../components/ExpenseList";

class UsersList extends Component {
  constructor() {
    super();
    this.state = {
      name: "",
      amount: "",
      description: "",
      expenses: []
    };
  }

  //calculates total amount spent on each render
  calculateSum = () => {
    const { expenses } = this.state;
    const total = expenses.reduce((a, b) => a + (parseInt(b.amount) || 0), 0);
    return total;
  };

  // adds single row for an individual expense
  addExpense = event => {
    event.preventDefault();
    this.setState({
      expenses: [
        ...this.state.expenses,
        {
          name: this.state.name,
          amount: this.state.amount,
          description: this.state.description
        }
      ],
      name: "",
      amount: "",
      description: "",
      isResult: false
    });
  };

  // shows the SettleResults component with split Bill
  splitExpense = () => {
    this.setState({
      isResult: true
    });
  };

  //removes the row onClick of remove button
  removeItem = indx => {
    this.setState(prevState => {
      const { expenses } = prevState;
      expenses.splice(indx, 1);
      return { ...prevState, expenses, isResult: false };
    });
  };

  // handles input changes for user entered fields
  handleInputChange = event => {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      [name]: target.type === "number" ? parseInt(value) : value
    });
  };

  render() {
    const total = this.calculateSum();
    const { isResult, expenses, name, amount, description } = this.state;
    return (
      <div className="content">
        <div className="container">
          <form className="row" onSubmit={this.addExpense}>
            <div className="col-xs-12 col-lg-3">
              <input
                type="text"
                name="name"
                className="form-control"
                value={name}
                required
                placeholder="Who Paid"
                onChange={this.handleInputChange}
              />
            </div>
            <div className="col-xs-12 col-lg-3">
              <input
                type="number"
                name="amount"
                value={amount}
                className="form-control"
                placeholder="Amount"
                required
                onChange={this.handleInputChange}
              />
            </div>
            <div className="col-xs-12 col-lg-3">
              <input
                type="text"
                name="description"
                value={description}
                className="form-control"
                placeholder="Description"
                onChange={this.handleInputChange}
              />
            </div>
            <button type="submit" className="btn btn-primary action-wrapper btn-primary" disabled={!name}>
              {ADD_EXPENSE}
            </button>
          </form>

          {expenses && expenses.length ? (
            <ExpenseList
              total={total}
              expenses={expenses}
              splitExpense={this.splitExpense}
              removeItem={this.removeItem}
            />
          ) : null}

          {isResult && <SettleResults expenses={expenses} totalAmount={total} />}
        </div>
      </div>
    );
  }
}

export default UsersList;
