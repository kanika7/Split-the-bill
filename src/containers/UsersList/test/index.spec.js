import React from "react";
import { shallow } from "enzyme";
import UsersList from "../index";
import SettleResults from "../../SettleResults";

describe("<UsersList /> Test Suits", () => {
  let wrapper;
  let instance;
  function renderShallow() {
    wrapper = shallow(<UsersList />);
    instance = wrapper.instance();
  }

  it("renders snapshot", () => {
    renderShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("handle input values change by calling handleInputChange", () => {
    let event = {
      target: {
        name: "name",
        value: "abc"
      },
      preventDefault: () => {}
    };
    renderShallow();
    wrapper
      .find(".form-control")
      .first()
      .simulate("click");
    instance.handleInputChange(event);
    expect(wrapper.state()).toStrictEqual({
      amount: "",
      description: "",
      expenses: [],
      name: "abc"
    });

    event = {
      target: {
        name: "amount",
        value: "200"
      },
      preventDefault: () => {}
    };
    renderShallow();
    wrapper
      .find(".form-control")
      .at(2)
      .simulate("click");
    instance.handleInputChange(event);
    expect(wrapper.state()).toStrictEqual({
      amount: "200",
      description: "",
      expenses: [],
      name: ""
    });

    event = {
      target: {
        name: "description",
        value: "test description"
      },
      preventDefault: () => {}
    };
    renderShallow();
    wrapper
      .find(".form-control")
      .last()
      .simulate("click");
    instance.handleInputChange(event);
    expect(wrapper.state()).toStrictEqual({
      amount: "",
      description: "test description",
      expenses: [],
      name: ""
    });
  });

  it("handles addExpense to add individual expenses", () => {
    const event = {
      preventDefault: () => {}
    };
    wrapper.setState({
      name: "abc",
      amount: 20,
      description: "test desc"
    });
    wrapper.update();
    instance.addExpense(event);
    expect(wrapper.state()).toStrictEqual({
      expenses: [
        {
          name: "abc",
          amount: 20,
          description: "test desc"
        }
      ],
      name: "",
      amount: "",
      description: "",
      isResult: false
    });
  });

  it("handles removeItem and removes item from expenses array", () => {
    renderShallow();
    wrapper.setState({
      expenses: [
        {
          name: "abc",
          amount: 20,
          description: "test desc1"
        },
        {
          name: "def",
          amount: 50,
          description: "test desc2"
        }
      ]
    });
    // wrapper.update();
    jest.spyOn(instance, "removeItem");
    instance.removeItem(0);
    expect(wrapper.state()).toStrictEqual({
      expenses: [
        {
          name: "def",
          amount: 50,
          description: "test desc2"
        }
      ],
      isResult: false,
      name: "",
      amount: "",
      description: ""
    });
  });

  it("handles splitExpense and updates State to show Split Bill", () => {
    renderShallow();
    jest.spyOn(instance, "splitExpense");
    instance.splitExpense();
    expect(wrapper.state()).toStrictEqual({
      expenses: [],
      isResult: true,
      name: "",
      amount: "",
      description: ""
    });
    expect(<SettleResults />).toBeDefined();
  });
});
