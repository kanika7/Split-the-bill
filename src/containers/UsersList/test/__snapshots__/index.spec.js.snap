// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsersList /> Test Suits renders snapshot 1`] = `
<div
  className="content"
>
  <div
    className="container"
  >
    <form
      className="row"
      onSubmit={[Function]}
    >
      <div
        className="col-xs-12 col-lg-3"
      >
        <input
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Who Paid"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="col-xs-12 col-lg-3"
      >
        <input
          className="form-control"
          name="amount"
          onChange={[Function]}
          placeholder="Amount"
          required={true}
          type="number"
          value=""
        />
      </div>
      <div
        className="col-xs-12 col-lg-3"
      >
        <input
          className="form-control"
          name="description"
          onChange={[Function]}
          placeholder="Description"
          type="text"
          value=""
        />
      </div>
      <button
        className="btn btn-primary action-wrapper btn-primary"
        disabled={true}
        type="submit"
      >
        Add Expense
      </button>
    </form>
  </div>
</div>
`;
